repository.py

```diff
@@ -130,14 +130,15 @@ def _parse_repodata(self, accessor):
         repomdfp.close()
 
         primaryfp = accessor.openAddress(primary_location)
-        # Open compressed xml using cpiofile._Stream which is an adapter between CpioFile and a stream-like object.
-        # Useful when specifying the URL for HTTP or FTP repository - A simple GzipFile object will not work in this situation.
-        primary_xml = cpiofile._Stream("", "r", "gz", primaryfp, 20*512)
-        primary_dom = parse(primary_xml)
-        package_names = primary_dom.getElementsByTagName("location")
-        package_sizes = primary_dom.getElementsByTagName("size")
-        package_checksums = primary_dom.getElementsByTagName("checksum")
-        primary_xml.close()
+
+        # HTTP/FTP accessors don't implement tell(), so read the data into a buffer
+        with BytesIO(primaryfp.read()) as fileobj:
+            with gzip.GzipFile("", "r", fileobj=fileobj) as primary_xml:
```

Comment:
All support versions of python support multiple context managers in a singe `with` statement, so you should be able to write:
```python
with BytesIO(primaryfp.read()) as fileobj, \
     gzip.GzipFile("", "r", fileobj=fileobj) as primary_xml:
    dom = parse(...
```

Reply:
I can confirm that this works

Reply:
Thanks, applied with shortening of the internal variables in the context:
- The temporary variables inside the context management can use easy short names
- This makes them more natural to read and understand, as the context is small
- This allows to fit the context managers on a line. It makes it even clearer to read:
```py
# HTTP&FTP accessors don't implement tell(): read xml using BytesIO to gunzip it
with BytesIO(primaryfp.read()) as fp, gzip.GzipFile("", "r", fileobj=fp) as xml:
    dom = parse(xml)
    package_names = dom.getElementsByTagName("location")
    package_sizes = dom.getElementsByTagName("size")
    package_checksums = dom.getElementsByTagName("checksum")
```
Closing comments:
- The shortened variables allow the line to fit into 88 columns
- A width of 88 columns is the default width of black
- With this, this commit is also black-compliant.
  - If we choose to reformat using black or darker (only changed lines) this is ready for it.
Final remark:
- I don't like the `\` notation because:
  - It breaks the flow of reading a statement into two lines that have to be put together.
  - It is not uses by black/darker to break long lines,
    - `black` would end up reformat it in this way to remove the `\`:
```py 
with BytesIO(primaryfp.read()) as fp, gzip.GzipFile(
   "", "r", fileobj=fp
) as primary_xml:
    dom = parse(primary_xml)
    package_names = dom.getElementsByTagName("location")
    package_sizes = dom.getElementsByTagName("size")
    package_checksums = dom.getElementsByTagName("checksum")

```
The short form that I've come up now, is much more readable than both other options to me,
because it almost reads like a sentence (I don't like multiline statements because of that),
and find this to be much more (human)-language-like to read:
```py
with BytesIO(primaryfp.read()) as fp, gzip.GzipFile("", "r", fileobj=fp) as xml:
    dom = parse(xml)
    package_names = dom.getElementsByTagName("location")
    package_sizes = dom.getElementsByTagName("size")
    package_checksums = dom.getElementsByTagName("checksum")

```
Of overall, that's a nice improvement